### PR TITLE
Add Modulated Deformable Convolution layer

### LINKED ISSUE #179: Deformable Convolutional Op Support
**System information**
- TensorFlow version (you are using):2.0 alpha
- TensorFlow Addons version:0,21
- Is it in the tf.contrib (if so, where):no
- Are you willing to contribute it (yes/no):yes,if i know how to do that
- Are you willing to maintain it going forward? (yes/no):yes

**Describe the feature and the current behavior/state.**
Deformable Convolutional Network(https://arxiv.org/pdf/1703.06211.pdf,https://arxiv.org/pdf/1703.06211.pdf) has achieved good results in the du detection task.But the current version of Tensorflow does not seem to implement deformable convolution operations.The efficiency of some third-party implementations of deformable convolution operations is lower than that of the mxnet version(https://github.com/msracver/Deformable-ConvNets).I hope that Tensorflow can add native deformable convolution operations in future versions.
Thanks.
**Will this change the current api? How?**
Yes,Add a new op
**Who will benefit with this feature?**
Any one want to use tensorflow for object detection.
**Any Other info.**

### PR DESCRIPTION
**Pull request by Licht-T**
# Description

This is the implementation of [Modulated Deformable Convolution](https://arxiv.org/abs/1811.11168). This fixes #179. There is another PR #1129, but that is stale and contaminated by [unknown license codes](https://github.com/tensorflow/addons/pull/1129#issuecomment-662063946). So I re-implemented this which is based on [the Torchvision's implementation](https://github.com/pytorch/vision/pull/1586).

~~I knew this would be too big to get reviewed, so, at first, I made the CPU only kernel with Eigen. When the CPU kernel is good enough, I will move to the next step: the GPU kernel implementation.~~
**UPDATE: GPU kernel available!**

## Type of change

- [ ] Bug fix
- [ ] New Tutorial
- [ ] Updated or additional documentation
- [ ] Additional Testing
- [ ] New Activation and the changes conform to the [activation contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/activations/README.md#contribution-guidelines)
- [ ] New Callback and the changes conform to the [callback contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/callbacks/README.md#contribution-guidelines)
- [ ] New Image addition and the changes conform to the [image op contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/image/README.md#contribution-guidelines)
- [x] New Layer and the changes conform to the [layer contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/layers/README.md#contribution-guidelines)
- [ ] New Loss and the changes conform to the [loss contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/losses/README.md#contribution-guidelines)
- [ ] New Metric and the changes conform to the [metric contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/metrics/README.md#contribution-guidelines)
- [ ] New Optimizer and the changes conform to the [optimizer contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/optimizers/README.md#contribution-guidelines)
- [ ] New RNN Cell and the changes conform to the [rnn contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/rnn/README.md#contribution-guidelines)
- [ ] New Seq2seq addition and the changes conform to the [seq2seq contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/seq2seq/README.md#contribution-guidelines)
- [ ] New Text addition and the changes conform to the [text op contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/text/README.md#contribution-guidelines)

# Checklist:

- [x] I've properly [formatted my code according to the guidelines](https://github.com/tensorflow/addons/blob/master/CONTRIBUTING.md#coding-style)
    - [ ] By running Black + Flake8
    - [x] By running pre-commit hooks
- [x] This PR addresses an already submitted issue for TensorFlow Addons
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] This PR contains modifications to C++ custom-ops

# How Has This Been Tested?
`pytest` unit-testing

### CODE REVIEW

**Comment by Licht-T** [comment]
I signed CLA.

**Comment by Licht-T** [comment]
Compliant test fails due to [`conv_utils`](https://github.com/tensorflow/addons/pull/2196/files#diff-e2e884b1ede01201987546d2c8797720R22). It seems that there is no alternative on public TensorFlow Python API.

**Comment by Licht-T** [comment]
I found that Addons has `tensorflow_addons.utils.keras_utils` and moved some private APIs into it.

**Comment by Licht-T** [comment]
All green! Ready for review!

**Comment by Licht-T** [comment]
I couldn't wait to make the GPU kernel until the CPU one reviewed!

**Comment by bhack** [comment]
I put this under ecosystem review. If any project in TF ecosystem is interested in this or has this in It own internal roadmap we will try to review It here.

**Comment by Licht-T** [comment]
@bhack Great! Thanks!

**Comment by Licht-T** [comment]
Just bump not to be forgotten.

**Comment by Licht-T** [comment]
@bhack Hi, how is the review going?

**Comment by tomerk** [comment]
Checked in w/ @ematejska. The internal TF API owners & oss outreach will start a regular (bi-weekly?) review of PRs marked w/ the ecosystem-review label.

**Comment by Licht-T** [comment]
Hey, how is this going?

**Comment by tomerk** [comment]
Thanks for the ping! Setting up the process has been a bit slow, but our first meeting to review the ecosystem-review labeled PRs will be tomorrow, so I should have more info soon :)

**Comment by tomerk** [comment]
Notes from ecosystem review: If this was more mainstream it would be a good fit for KerasCV, but for now we think addons makes sense (this layer is not currently on tf's roadmap)

**Comment by Licht-T** [comment]
Thanks for your eco-system review!

What should I do next? Any requests are welcome since I know this is quite massive!

**Comment by bhack** [comment]
> Thanks for your eco-system review!
> 
> What should I do next? Any requests are welcome since I know this is quite massive!

Can you split your contribution in more than one PR? Or  is this just a single component PR?

**Comment by cheng-yu-td** [comment]
> > Thanks for your eco-system review!
> > What should I do next? Any requests are welcome since I know this is quite massive!
> 
> Can you split your contribution in more than one PR? Or is this just a single component PR?

This is one single feature, deformable convolution layer. I don't see how this can be split into multiple pull requests.

**Comment by cheng-yu-td** [comment]
Thanks @Licht-T  I look forward to this feature! Can't wait to use this for experiments.

**Comment by bhack** [comment]
This PR Is now near 3k lines. 

Have we already evaluated the performance gap of a compositional (also if v1) implementation like https://tensorlayer.readthedocs.io/en/latest/_modules/tensorlayer/layers/convolution/deformable_conv.html ?

File: tensorflow_addons/layers/deformable_conv2d.py [hunk 1]
```diff
@@ -0,0 +1,340 @@
...
+        self.kernel_regularizer = tf.keras.regularizers.get(kernel_regularizer)
+        self.bias_regularizer = tf.keras.regularizers.get(bias_regularizer)
+        self.kernel_constraint = tf.keras.constraints.get(kernel_constraint)
+        self.bias_constraint = tf.keras.constraints.get(bias_constraint)
+
+        if self.padding == "causal":
+            raise ValueError("Causal padding is not supported.")
+
+        if self.data_format != "channels_first":
+            raise ValueError("`channels_last` data format is not supported.")
```
Comment (anshkumar) [inline tensorflow_addons/layers/deformable_conv2d.py:183]: "channels_last" is the default data_format in tf.keras.layers.Conv2D. If user is selecting "channels_last", the output should be transposed, instead of raising error.

**Comment by Cospel** [comment]
The code looks great, we are using it successfully in one of our projects. Is anybody still working on it? It's been more than a year of waiting on deformable conv :) ...

**Comment by Licht-T** [comment]
Thanks for a lot of comments. I'm back!

Now I have much time to focus on this PR and will resolve some conflicts this week.

File: tensorflow_addons/layers/deformable_conv2d.py (hunk 1 above)
Comment (Licht-T) [reply tensorflow_addons/layers/deformable_conv2d.py:183]: `swapaxis` makes a lot of computing cost. IMO, `swapaxis` should not be implicitly applied.

**Comment by Licht-T** [comment]
That's difficult as `cheng-yu-td` told, but I should do so if I can. Any ideas are welcome.
> Can you split your contribution in more than one PR? Or is this just a single component PR?

**Comment by Licht-T** [comment]
@bhack Not tested yet. I'll try.

According to https://github.com/tensorlayer/tensorlayer/issues/641#issuecomment-391239510, TensorLayer implementation is 100 times slower than the original MXNet implementation. That's why we need the optimized C++ code for Deformable Convolution.
> Have we already evaluated the performance gap of a compositional (also if v1) implementation like https://tensorlayer.readthedocs.io/en/latest/_modules/tensorlayer/layers/convolution/deformable_conv.html ?

**Comment by bhack** [comment]
> According to tensorlayer/tensorlayer#641 (comment)

It is 3 years old. We could check what happens now with the current compile stack.


As a general side note:

/cc @yarri-oss @tomerk  As custom kernels are going to create a quite large code and maintainership overhead in the repo can we have a contact point in the XLA/HLO team to just undertestand if and when we could achieve enought performance with a compositional/compiling path (e.g. trigger some FR to the compiler stack)? 

We really want reduce the number of maitnained custom ops in the library.

See also https://discuss.tensorflow.org/t/tfr-compositional-ops/

**Comment by Licht-T** [comment]
@bhack I did some tests. In the best case, this PR is approx. 1000 times faster than the TensorLayer one.

- Input height/width v.s. Runtime plot with fixed batch size: [perf_hw.py](https://gist.github.com/Licht-T/03b104eb9c918fdd2c8ee37cbdade9e1)
perf_hw

- Batch size v.s. Runtime plot with fixed input width/height: [perf_batch.py](https://gist.github.com/Licht-T/1eacde305f52af9fbcb7d71b8229b80a)
perf_batch

**Comment by Licht-T** [comment]
I know this PR is quite a large code. I don't want to make maintainers annoyed by the large custom ops code, but I just want to provide the Deformable Convolution, which is one of the greatest achievements in CNN research, to TensorFlow users. 

If we can achieve good performance in another way like XLA/HLO, I should/will re-implement this PR in that way.

**Comment by bhack** [comment]
@Licht-T Yes thank you for effort.
The main issue is that Codeowners on average disappaer quite fast see the [last two graphs here](https://cauldron.io/project/4134?from_date=2020-06-12&to_date=2021-06-12&tab=community-reviews) and https://github.com/tensorflow/addons/pull/2024). 

In this specific case we will have near 3k lines to maintain. 
It is  I've asked to @seanpmorgan @tomerk @yarri-oss if we could have some feedback by the compiler team about compositional alternatives in cases like this.

Also I think that you need to test the tensorlayer implementation with `jit_compile` in core functions to test it with XLA (or duplicate the class yourself to set core parts under XLA).

You can inspect the compiled function with :

https://www.tensorflow.org/xla#inspect_compiled_programs


If you could profile it could be also interesting to see if there is any specific dominant ops.

https://www.tensorflow.org/tensorboard/tensorboard_profiling_keras

You can use profiler start and stop if you don't want to use callbacks
 `https://www.tensorflow.org/api_docs/python/tf/profiler/experimental/start?hl=en`

**Comment by peter-kettmann** [comment]
> @bhack Thanks for your comments! I did some tests. In the best case, this PR is approx. 1000 times faster than the TensorLayer one.
> 
> * Input height/width v.s. Runtime plot with fixed batch size: [perf_hw.py](https://gist.github.com/Licht-T/03b104eb9c918fdd2c8ee37cbdade9e1)
> 
> perf_hw
> 
> * Batch size v.s. Runtime plot with fixed input width/height: [perf_batch.py](https://gist.github.com/Licht-T/1eacde305f52af9fbcb7d71b8229b80a)
> 
> perf_batch

That looks really great! We are currently using deformable convolutions for a project and are using the tensorflowlayer version. But it's too slow for our use case. So I tried your implementation but ran into problems. How do I correctly build the code including the deformable conv layer?

So far I cloned your repo, switched to add-deformable-conv branch (tried the add-deformable-conv-internal as well) and followed the instructions to install tensorflow_addons from source. The build process runs fine (just some bazel warnings). However, installing the wheel and trying a simple code

```
import tensorflow as tf
import tensorflow_addons

input_layer = tf.keras.layers.Input(shape=(512,512,3))
channels_first = tf.transpose(input_layer, (0,3,1,2))

offset_1 = tf.keras.layers.Conv2D(2, (3,3), padding='same', data_format='channels_first', kernel_initializer='zeros')(channels_first)
deformable_1 = tensorflow_addons.layers.DeformableConv2D(3, (1,1), padding='same')([channels_first, offset_1])

channels_last = tf.transpose(deformable_1, (0,2,3,1))
model = tf.keras.Model(inputs=input_layer, outputs=channels_last)

test_data = tf.random.uniform((32,512,512,3), dtype=tf.float32)

out = model(test_data)
```
throws this error:

`NotFoundError: /net/io-truenas/mnt/pool0/workspaces/peter/code/addons/venv/lib/python3.6/site-packages/tensorflow_addons/custom_ops/layers/_deformable_conv2d_ops.so: undefined symbol: _ZN10tensorflow7strings8internal9CatPiecesB5cxx11ESt16initializer_listIN4absl14lts_2020_09_2311string_viewEE`

What am I doing wrong? Some help would be much appreciated :) 

Some additional info: I'm using

Ubuntu 18.04
Python 3.6.9
tensorflow 2.5.0
bazel 3.7.2
CUDA 11.2
CUDNN 8.2

and the warnings are several

`/usr/local/cuda-11.2/bin/../targets/x86_64-linux/include/thrust/detail/config/cpp_dialect.h:118:13: warning: Thrust requires C++14. Please pass -std=c++14 to your compiler. Define THRUST_IGNORE_DEPRECATED_CPP_DIALECT to suppress this message. THRUST_COMPILER_DEPRECATION(C++14, pass -std=c++14 to your compiler);`

and a

`WARNING: home/peter/.cache/bazel/_bazel_peter/c6b879c04b9dc802e9234eb6fb8e9fc9/external/local_config_tf/BUILD:12633:8: target 'libtensorflow_framework.so.2' is both a rule and a file; please choose another name for the rule`

**Comment by axeldavy** [comment]
Hey, I've tried this deformable convolution implementation in my project, and it seems to work as expected.

However the performance hit compared to a normal convolution is huge.
Indeed the implementation doesn't support the NHWC ordering, which means I have to swap before and after the convolution, and it addition it doesn't support float16 or mixed_float16, and thus doesn't use fp16 tensorcores.

Would it be possible to support NHWC ordering and mixed_fp16/fp16 ?

An additional thought: unless I'm mistaken, it seems that the modulated convolution could be implemented as first a layer that grabs the values at the predicted positions (with bilinear interpolation), and applies the mask, and concatenates the values (for example for a 3x3 convolution the number of channels of the output would be multiplied by 9), then a standard 1x1 convolution.
Possibly it would be easier to implement and maintain that way (only the first layer needs to be implemented).

**Comment by anshkumar** [comment]
@Licht-T I tried compiling it using cuda 11.6, cudnn 8.3 but getting following error on usage:

```
2022-02-23 16:37:03.133656: F tensorflow_addons/custom_ops/layers/cc/kernels/deformable_conv2d_op_gpu.cu.cc:292] Non-OK-status: GpuLaunchKernel(DeformableIm2ColKernel<float>, config.block_count, config.thread_per_block, 0, device.stream(), b, num_kernels, p, input_eigen_tensor, offset_eigen_tensor, mask_eigen_tensor, column_buffer_eigen_tensor) status: Internal: too many resources requested for launch
```

**Comment by anshkumar** [comment]
> @Licht-T I tried compiling it using cuda 11.6, cudnn 8.3 but getting following error on usage:
> 
> ```
> 2022-02-23 16:37:03.133656: F tensorflow_addons/custom_ops/layers/cc/kernels/deformable_conv2d_op_gpu.cu.cc:292] Non-OK-status: GpuLaunchKernel(DeformableIm2ColKernel<float>, config.block_count, config.thread_per_block, 0, device.stream(), b, num_kernels, p, input_eigen_tensor, offset_eigen_tensor, mask_eigen_tensor, column_buffer_eigen_tensor) status: Internal: too many resources requested for launch
> ```

Changing every occurrence of `config.thread_per_block` with `config.thread_per_block/2` fixed this.

**Comment by SimonBiggs** [comment]
Hi @bhack,

If I was to give this PR another try but attempt to split it down into smaller more bite sized PRs would you/others be willing to review and merge?

Another caveat, is I would want to have my initial "hello world" implementation be a 3D deformable convolution as opposed to 2D within this PR.

Although I can't prove that I'll stick around after the PR, I do maintain [a package in my field](https://github.com/pymedphys/pymedphys) which pulls in outside contributions, and by feeling some of the pain of having code-owners disappear, I hope my word has enough weight when I say I want to stick around and help when there is future work/issues/PRs going on around the code that I contribute.

Cheers,
Simon

**Comment by SimonBiggs** [comment]
> An additional thought: unless I'm mistaken, it seems that the modulated convolution could be implemented as first a layer that grabs the values at the predicted positions (with bilinear interpolation), and applies the mask, and concatenates the values (for example for a 3x3 convolution the number of channels of the output would be multiplied by 9), then a standard 1x1 convolution.
> Possibly it would be easier to implement and maintain that way (only the first layer needs to be implemented).

@axeldavy, if I was to take this on, might you want to help me?

**Comment by bhack** [comment]
@SimonBiggs Can you open a ticket to propose this in Keras-CV https://github.com/keras-team/keras-cv/


As many of the computer vision related contributions are converging there.

**Comment by SimonBiggs** [comment]
I notice the following comment re custom ops:

https://github.com/keras-team/keras-cv/blob/master/.github/CONTRIBUTING.md#contributing-custom-ops

Would that make this contribution the first custom op in Keras-CV that pulls in CUDA code?

I would aim to use the approach that @axeldavy mentioned, so the custom op section would hopefully have a reasonably smaller surface area.

**Comment by bhack** [comment]
We could start with a ticket there to discuss the topic and how to cover this convolution before creating a PR.

**Comment by SimonBiggs** [comment]
Thanks @bhack,

New issue now over at https://github.com/keras-team/keras-cv/issues/1140.

Cheers :),
Simon
